lib/Serialization/Deserialization.cpp

```diff
@@ -275,7 +293,7 @@ Expected<Pattern *> ModuleFile::readPattern(DeclContext *owningDC) {
   llvm::BitstreamEntry next =
       fatalIfUnexpected(DeclTypeCursor.advance(AF_DontPopBlockAtEnd));
   if (next.Kind != llvm::BitstreamEntry::Record)
-    fatal();
+    return diagnoseFatal();
```

Comment:
These low-level failures happen only during development or when the swiftmodule being read is corrupt. IMO even LLDB should fail here. If the swiftmodule is corrupt continuing compilation is at risk to use pointers to random data. The custom error object may help here.

Reply:
Terminating the debugger is always undesirable for the user because they will loose all the state accumulated in all of their debug sessions. DiagnoseFatal sets the ASTContext into a fatal error state, which is a signal for LLDB to not use this context any more. LLDB will then proceed to tear down this ASTContext only and create a new one in its place. If the compiler actually manages to crash between the tear down and the fatal error as a result of the deserialization failure, we won't be any worse off than before from the end-user's perspective.